omen2/object.py

```diff
@@ -0,0 +1,252 @@
+
+    def __hash__(self):
+        return hash(self._to_pk_tuple())
+
+    def _to_pk_tuple(self):
+        return tuple(sorted(self._to_pk().items()))
+
+    def __lt__(self, other: "ObjBase"):
+        return self._to_pk_tuple() < other._to_pk_tuple()
+
+    def __init_subclass__(cls, **_kws):
+        # you must set these in the base class
+        assert cls._pk, "All classes must have a _pk"
+
+    def __init__(self, **kws):
+        """Override this to control initialization, generally calling it *after* you do your own init."""
+        self._meta = ObjMeta()
+        self._meta.lock = RLock()
+        self._check_kws(kws)
+        self._meta.new = True
+
+    def _save_pk(self):
+        self._meta.pk = self._to_pk()
+        self._meta.restore = self._to_db()
+
+    @classmethod
+    def _from_db(cls, dct):
+        """Override this if you want to change how deserialization works."""
+        return cls(**dct)
+
+    @classmethod
+    def _from_db_not_new(cls, dct):
+        """Override this if you want to change how deserialization works."""
+        ret = cls._from_db(dct)
+        ret._meta.new = False
+        ret._save_pk()
+        return ret
+
+    def _check_kws(self, dct):
+        for k in dct:
+            if k not in self.__dict__:
+                raise AttributeError(
+                    "%s not a known attribute of %s" % (k, self.__class__.__name__)
+                )
+
+    def _matches(self, dct):
+        for k, v in dct.items():
+            if getattr(self, k) != v:
+                return False
+        return True
+
+    def _update(self, dct):
+        self.__dict__.update(dct)
+
+    def _bind(self, table: "Table" = None, manager: "Omen" = None):
+        if table is None:
+            table = manager[self._table_type]
+        self._meta.table = table
+
+    def _to_db(self):
+        """Get dict of serialized data from self."""
+        ret = {}
+        for k in self._table_type.field_names:
+            v = getattr(self, k)
+            if v is None:
+                continue
+            if hasattr(v, "_to_db"):
+                # pylint: disable=no-member
+                v = v._to_db()
+            ret[k] = v
+        return ret
+
+    def _to_pk(self):
+        """Get dict of serialized data from self, but pk elements only.
+
+        You will have to override this if you're overriding _to_db, and
+        any of the transformed values are part of your primary key.
+
+        Guaranteed compatible override would be:
+
+        def _to_pk(self):
+            dct = self._to_db()
+            return {k: dct[v] for k in self._pk}
+
+        The only reason this isn't used by default is efficiency.
+        """
+        ret = {}
+        for k in self._pk:
+            v = getattr(self, k)
+            if v is None:
+                raise OmenNoPkError("invalid primary key")
+            ret[k] = v
+        return ret
+
+    def _need_id(self):
+        need_id_field = None
+        for field in self._pk:
```

Comment:
if the table has multiple pks, we will arbitrarily autoinc one of them?

Reply:
it should fail completely.